File: LeetCode 77 39 40-Combination and Combination sum/combinations_duplicate.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func combine(elements []int, k int) [][]int {
	sort.Ints(elements)
	combinations := [][]int{}
	backtrack(elements, k, 0, []int{}, &combinations)

	return combinations
}

func backtrack(elements []int, k int, pos int, currentCombination []int, combinations *[][]int) {
	if len(currentCombination) == k {
		// deep copy
		newCombination := append([]int{}, currentCombination...)
		fmt.Println("new combination:", newCombination)
		*combinations = append(*combinations, newCombination)
		return
	}

	prev := math.MaxInt32
	for i := pos; i < len(elements); i++ {
		currentElement := elements[i]
		if currentElement == prev {
			fmt.Printf("skip %d\n", currentElement)
			continue
		}
		currentCombination = append(currentCombination, currentElement)
		backtrack(elements, k, i+1, currentCombination, combinations)
		// clean
		currentCombination = currentCombination[:len(currentCombination)-1]
		prev = currentElement
	}
}

func main() {
	elements := []int{1, 2, 3, 2}
	k := 3
	fmt.Println("combinations:", combine(elements, k))
}

/*
outout:
new combination: [1 2 2]
new combination: [1 2 3]
skip 2
new combination: [2 2 3]
skip 2
combinations: [[1 2 2] [1 2 3] [2 2 3]]
*/

File: LeetCode 77 39 40-Combination and Combination sum/combinations_duplicate.py
class Solution:
    def combine(self, elements, k):
        elements.sort()
        combinations = []
        
        def backtrack(pos, currentCombination):
            if len(currentCombination) == k:
                # copy and append
                print("new combination", currentCombination)
                combinations.append(currentCombination[::])
                return
            
            prev = None
            for i in range(pos, len(elements)):
                currentElement = elements[i]
                if prev == currentElement:
                    print("skip", currentElement)
                    continue
                currentCombination.append(currentElement)
                backtrack(i+1, currentCombination)
                # clean
                currentCombination.pop()
                prev = currentElement
            
        backtrack(0, [])
        return combinations


if __name__ == "__main__":
    solution = Solution()
    elements = [1, 2, 3, 2]
    k = 3
    print("combinations", solution.combine(elements, k))


"""
output:
('new combination', [1, 2, 2])
('new combination', [1, 2, 3])
('skip', 2)
('new combination', [2, 2, 3])
('skip', 2)
('combinations', [[1, 2, 2], [1, 2, 3], [2, 2, 3]])
"""